docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,754 @@
+        size_t* evidence_buffer_size,
+        uint8_t** endorsements_buffer,
+        size_t* endorsements_buffer_size);
+
+    /**
+     * Frees the generated attestation evidence.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] evidence_buffer A pointer to the evidence buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_evidence)(
+        oe_attestation_plugin_t* plugin_context,
+        uint8_t* evidence_buffer);
+
+    /**
+     * Frees the generated attestation endorsements.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] endorsements_buffer A pointer to the endorsements buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_endorsements)(
+        oe_attestation_plugin_t* plugin_context,
+        uint8_t* endorsements_buffer);
+
+    /**
+     * Verifies the attestation evidence and returns the claims contained in
+     * the evidence.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] evidence_buffer The evidence buffer.
+     * @param[in] evidence_buffer_size The size of evidence_buffer in bytes.
+     * @param[in] endorsements_buffer The endorsements buffer.
+     * @param[in] endorsements_buffer_size The size of endorsements_buffer in bytes.
+     * @param[in] input_validation_time Optional datetime to use when verifying
+     * evidence. If not specified, it will use the creation_datetime of the
+     * endorsements (if any endorsements are provided).
+     * @param[out] claims The list of claims.
+     * @param[out] claims_length The number of claims.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*verify_evidence)(
+        oe_attestation_plugin_t* plugin_context,
+        const uint8_t* evidence_buffer,
+        size_t evidence_buffer_size,
+        const uint8_t* endorsements_buffer,
+        size_t endorsements_buffer_size,
+        const oe_datetime_t* input_validation_time,
+        oe_claim_t** claims,
+        size_t* claims_length);
+};
+```
+
+Here is the rationale for each element in the plugin struct:
+
+- `format_id`
+  - Each plugin needs an unique identifier to distinguish itself.
+- `on_register` and `on_unregister`
+  - A plugin might require some setup or teardown when it is registered or
+    unregistered, so these functions are required. Furthermore, a plugin
+    might require configuration, which is why there is a `config_data`
+    parameter. The configuration data can be plugin specific, so no format is
+    specified in this proposal.
+- `get_evidence`, `free_evidence`, and `free_endorsements`
+  - Producing evidence and endorsements is necessary for attestation.
+  - `flags` field to determine local vs. remote attestation.
+  - There is a `custom_claims` parameter because many attestation protocols
+    require the enclave to sign some claim from a relying party. For example,
```

Comment:
suggest "some claim" -> "some data", since a claim was defined as a name/value pair and it might not sign a name, just the value.